cpp/include/cudf/dictionary/dictionary_column_view.hpp

```diff
@@ -0,0 +1,235 @@
+                                           column_view const& dictionary,
+                                           rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+/**
+ * @brief Create a new dictionary column by removing any dictionary elements
+ * that are not referenced by any of the indices.
+ *
+ * The indices are updated to the new position values of the remaining dictionary elements.
+ *
+ * ```
+ * d1 = {["a","c","d"],[2,0,2,0]}
+ * d2 = remove_unused_dictionary(d1)
+ * d2 is now {["a","d"],[1,0,1,0]}
+ * ```
+ *
+ * @param dictionary_column Existing dictionary column.
+ * @param mr Resource for allocating memory for the output.
+ * @return New dictionary column.
+ */
+std::unique_ptr<column> remove_unused_dictionary( dictionary_column_view const& dictionary_column,
+                                                  rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+/**
+ * @brief Create a new dictionary column by applying only the specified dictionary elements
+ * to the existing dictionary_column.
+ *
+ * Any new elements found in the dictionary parameter are added to the output dictionary.
+ * Any existing dictionary elements not in the dictionary parameter are removed.
+ *
+ * The indices are update to reflect the position values of the new dictionary.
+ * Any indices pointing to removed elements are set to null.
+ *
+ * ```
+ * d1 = {["a","c","d"],[2,0,1,0]}
+ * d2 = set_dictionary(d1,["a","b","c"])
+ * d2 is now {["a","b","c"],[null,0,2,0]}
+ * ```
+ *
+ * @throw cudf_logic_error if the dictionary type does not match the dictionary type in
+ * the dictionary_column.
+ *
+ * @param dictionary_column Existing dictionary column.
+ * @param dictionary New keys to use for the output column.
+ * @param mr Resource for allocating memory for the output.
+ * @return New dictionary column.
+ */
+std::unique_ptr<column> set_dictionary( dictionary_column_view const& dictionary_column,
+                                        column_view const& dictionary,
+                                        rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+/**
+ * @brief Create a new dictionary column by merging the specified dictionary elements
+ * from two existing dictionary_columns.
+ *
+ * The indices are updated to appear as though the second dictionary_column is appended to the first.
+ *
+ * ```
+ * d1 = {["a","c","d"],[2,0,1,0]}
+ * d2 = {["b","e"],[1,0,1,1,0]}
+ * d3 = merge(d1,d2)
+ * d3 is now {["a","b","c","d","e"],[3,0,2,0,4,1,4,4,1]}
+ * ```
+ *
+ * @throw cudf_logic_error if the dictionary types do not match for both dictionary columns.
+ *
+ * @param dictionary_column Existing dictionary column.
+ * @param dictionary New keys to remove from the dictionary_column
+ * @param mr Resource for allocating memory for the output.
+ * @return New dictionary column.
+ */
+std::unique_ptr<column> merge_dictionaries( dictionary_column_view const& dictionary_column1,
```

Comment:
Is this function needed if `add_dictionary_keys` is flexible enough to support a dictionary column view as the second input? Can they be combined?

Reply:
The indices from the 2 dictionary column are also merged (appended).

Reply:
I think Mark's point is that the following are equivalent:
```c++
dictionary_column_view dcv1;
dictionary_column_view dcv2;

merged = merge_dictionaries(dcv1,dcv2);

// alternatively
merged_alt = add_keys(dcv1, dcv2.keys());
```

Reply:
These two are not the same. You may be confusing dictionaries with keys here.
`merge_dictionaries` will merge two dictionary columns -- produce new keys and append the indices from each dictionary column
`add_keys` simply adds the keys and adjusts the existing indices -- no new indices are added

Reply:
Aha, I see now. In my mind that's more of a concatenation operation than a merge.